libs/utils/analysis/frequency_analysis.py

```diff
@@ -110,6 +110,70 @@ def _dfg_cluster_frequency_residency(self, cluster, total=True):
             return residency.total
         return residency.active
 
+    def _dfg_frequency_transitions(self, entity):
+        """
+        Compute number of frequency transitions of a given entity. Entity can
+        be a CPU or a cluster name. Notice that we assume that frequency is
+        scaled at cluster level, therefore we always consider the first CPU of
+        a cluster for this computation.
+
+        Requires cpu_frequency events to be available in the trace.
+
+        :param entity: a CPU ID or a cluster name
+        :type entity: int or str
+
+        :returns: :mod:`pandas.DataFrame` - number of frequency transitions
+        """
+        if not self._trace.hasEvents('cpu_frequency'):
+            self._log.warn('Events [cpu_frequency] not found, '
+                           'cannot compute frequency transitions!')
+            return None
+
+        if isinstance(entity, str):
+            cpu = self._platform['clusters'][entity.lower()][0]
+        else:
+            cpu = entity
+
+        freq_df = self._dfg_trace_event('cpu_frequency')
+        cpu_freqs = freq_df[freq_df.cpu == cpu].frequency
+
+        # Remove possible duplicates (example: when devlib sets trace markers
+        # a cpu_frequency event is triggered that can generate a duplicate)
+        cpu_freqs = cpu_freqs.loc[cpu_freqs.shift(-1) != cpu_freqs]
+        transitions = cpu_freqs.value_counts()
+        # Convert frequencies to MHz
+        transitions.index = transitions.index / 1000
+        transitions.name = "transitions"
+        transitions.sort_index(inplace=True)
+        return pd.DataFrame(transitions)
+
+    def _dfg_frequency_transition_rate(self, entity):
+        """
+        Compute frequency transition rate of a given entity. Entity can
+        be a CPU or a cluster name. Notice that we assume that frequency is
+        scaled at cluster level, therefore we always consider the first CPU of
+        a cluster for this computation.
+
+        Requires cpu_frequency events to be available in the trace.
+
+        :param entity: a CPU ID or a cluster name
+        :type entity: int or str
+
+        :returns: :mod:`pandas.DataFrame - number of frequency transitions per
+            second
+        """
+        if not self._trace.hasEvents('cpu_frequency'):
```

Comment:
This check is already performed by the following call at line 170.

I would remove this one and make more generic the previous warn at line 127 by just saying something like:
```python
        if not self._trace.hasEvents('cpu_frequency'):
            self._log.warn('Events [cpu_frequency] not found, '
                           'frequency data not available')
```